### PR TITLE
Add support for bucket adaptation

### PR DESCRIPTION
**Pull request by mbasmanova**
Turn bucket adaptation into filter function and run it along with other filters.

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Comment by wenleix** [comment]
Curious if this is used to optimize for filter on `$bucket` column?

**Comment by mbasmanova** [comment]
@wenleix Curious if this is used to optimize for filter on $bucket column?

Bucket adaptation is used when reading data from older partitions that used larger buckets. In this case, we need to compute new buckets for each row and discard rows with the "wrong" bucket number.

**Comment by wenleix** [comment]
@mbasmanova I see. So basically this bucket adaption is not supported by `OrcSelectiveReader`.

I am wondering if it makes sense to run `HiveIntegrationSmokeTest` with aria scan enabled as well? -- This can be done in separate PR.  Because bucket adaption is also tested there 😄 . This also makes sure all known "basic query shapes" are tested ~

**Comment by mbasmanova** [comment]
@wenleix Wenlei, thanks for review.

> I am wondering if it makes sense to run HiveIntegrationSmokeTest with aria scan enabled as well?

We already do that. The test is TestHivePushdownIntegrationSmokeTest.

File: presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSourceFactory.java [hunk 1]
```diff
@@ -333,7 +344,14 @@ public static OrcSelectivePageSource createOrcPageSource(
                             hiveColumnIndex -> new VariableReferenceExpression(columnNames.get(hiveColumnIndex), columnTypes.get(hiveColumnIndex)),
                             hiveColumnIndex -> new InputReferenceExpression(inputs.get(hiveColumnIndex), columnTypes.get(hiveColumnIndex))));
 
-            List<FilterFunction> filterFunctions = toFilterFunctions(replaceExpression(remainingPredicate, variableToInput), session, rowExpressionService.getDeterminismEvaluator(), rowExpressionService.getPredicateCompiler());
+            Optional<BucketAdapter> bucketAdapter = bucketAdaptation.map(adaptation -> new BucketAdapter(
+                    Arrays.stream(adaptation.getBucketColumnIndices()).map(indexMapping::get).map(inputs::get).toArray(),
```
Comment (wenleix) [inline presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSourceFactory.java]: nit: Maybe add line breaks? such as:

```java
Arrays.stream(adaptation.getBucketColumnIndices())
              .map(indexMapping::get)
              .map(inputs::get)
              .toArray()
```

**Comment by mbasmanova** [comment]
@wenleix Wenlei, thanks for the review.

> Originally the bucket adaption is done in HivePageSource#getNextPage:

That's correct.

> Looks like this works well for OrcBatchPageSource . I am wondering why this mechanism somehow not work properly for OrcSelectivePageSource? (I think this is by design but just curious)

This could work for selective page source as well, but it would be very inefficient. Pushing this filter inside the selective reader allows the rows to be dropped early, e.g. before reading columns with no filter, and allowed to avoid data copying.

> I am raising this question because now it looks to me this part of "eligible row selection" can be applied twice: one inside OrcSelectivePageSource, and one is in HivePageSource?

Selective reader path doesn't use HivePageSource.

File: presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSourceFactory.java (hunk 1 above)
Comment (mbasmanova) [reply presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSourceFactory.java]: @wenleix Will do.

**Pull request merged**